IdentityCore/src/cache/metadata_cache/metadata/MSIDAuthorityMap.m

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#import "MSIDAuthorityMap.h"
+#import "MSIDAccountIdentifier.h"
+#import "MSIDCache.h"
+#import "MSIDAuthority.h"
+#import "MSIDAuthorityFactory.h"
+
+@implementation MSIDAuthorityMap
+{
+    MSIDCache *_authorityMap;
+}
+
+- (instancetype)initWithAccountIdentifier:(MSIDAccountIdentifier *)accountIdentifier
+                                 clientId:(NSString *)clientId
+{
+    if (!accountIdentifier || !clientId) return nil;
+    
+    self = [super init];
+    if (self)
+    {
+        self.accountIdentifier = accountIdentifier;
+        self.clientId = clientId;
+        _authorityMap = [MSIDCache new];
+    }
+    return self;
+}
+
+- (BOOL)addMappingWithRequestAuthority:(MSIDAuthority *)requestAuthority
+                     internalAuthority:(MSIDAuthority *)internalAuthority
+{
+    if (!internalAuthority.url.absoluteString || !internalAuthority.url.absoluteString) return NO;
+    
+    [_authorityMap setObject:internalAuthority.url.absoluteString forKey:requestAuthority.url.absoluteString];
+    return YES;
+}
+
+- (MSIDAuthority *)cacheLookupAuthorityForAuthority:(MSIDAuthority *)authority
+{
+    NSString *authorityStr = [_authorityMap objectForKey:authority.url.absoluteString];
+    
+    if (!authorityStr) return nil;
+    
+    NSURL *authorityUrl = [NSURL URLWithString:authorityStr];
+    
+    if (!authorityUrl) return nil;
+    
+    return [MSIDAuthorityFactory authorityFromUrl:authorityUrl context:nil error:nil];
```

Comment:
don't use authority factory. it uses magic strings and it won't work for b2c and other custom authorities

Reply:
It now just handles NSURL, not the authority object. It should be addressed in MSAL level.